### DISCUSSION TITLE
Resistor values for USB-UART 4.7K UPDI programming

### DISCUSSION THREAD

**Discussion by microPaul** [SpenceKonde/DxCore#94] [Q&A]
I've been looking at the voltage characteristics (high/low input and output voltage vs data sheet thresholds) and I'm thinking that this method of programming would get a lot more voltage margin if a Shottky diode was placed in parallel with the 4.7K resistor (cathode facing the USB-UART TX output),  I wrote up a rational for my thinking.  Please let me know if I'm looking at this the wrong way:  http://ad7i.net/main/analysis-of-resistor-values-for-pymcuprog-usbusart-serial-programming/

**Comment by SpenceKonde** [comment]
You are the second person to suggest a diode there to me in the past 24 hours. The other guy was an empiricist rather than a theoretician - he didn't have calculations, but he did claim to have achieved 600 kbaud. He also didn't have a series resistor in place I dont think, which probably isn;t ideal... (you might be fine on a tiny, where the pin driver on the UPDI pin is only slightly stronger than a typical pullup resistor... butcertainly non on an AVR DD-0series, where current is injected on reset for hv programming - at least that's what I've been told by someone who really ought to know. - and that would impoly that it is a full strength IO pin)

I think that should work, definitely... might need to be careful that the diode doesn't have an unduely long reverse recovery time... in general though, I would expect this to work and that most plausible values involved would improve performance

**Reply by microPaul** [reply]
Gosh, I didn't even actually test the diode idea in a circuit.  At least the empiricist did, so I'm glad someone tested it in practice.  But seriously, I do think a Schottky-D would be helpful, and perhaps allow for a larger value than 470 ohms, to further limit current when bad voltages are connected to UPDI.  As for the AVR-DD parts, I have no insight on that.  So if you have it from the horse's mouth so to speak that that may not work on the new DD parts, then of course that needs to be respected.

**Comment by SpenceKonde** [comment]
Without resistor in series, might be okay in tiny, almost certainly not on
DD. With resistor should be fine either way, and is a good idea regardless

____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore: Arduino support for almost every ATTiny microcontroller
Contact: ***@***.***

**Reply by microPaul** [reply]
Yes, I like the idea of a current limiting resistor, because too frequently when prototyping I connect things in a way that they should not be connected.  The more difficult it is to change out a part on a board the more likely it will be that that's the part that blows up..   On another note, in the next week or so I will go through the UPDI programming with the USB-UART adapter with an oscilloscope, to look at the various voltages with and without the Schottky diode when programming an AVR-DB as well as a ATtiny-1.  I'll report back what I find.

**Comment by SpenceKonde** [comment]
Yeah - any connector that can be connected bvackwards will be, guaranteed. If an expert is doing it, maybe only a few percent of the time, So everythung must be safe against that. Notice how the ISP header is in one of the few arrangements of those 6 pins that won't damage things backwards - or won't damage them quickly at least (the only low impedance path to positive voltage would be applied to the HV tolerant reset pin). FTDI serial? That's safe too! Vcc would be on target's RX (it's idle high, so it would look like adapter's idle TX line, no problem), gnd on whatever target expected DTR (which it would expect to be able to drive low) to be on... whether or not it was supposed to be supplying power, it's safer than about any other imaginable pinout. And in the case of Arduino, it's indefinitely safe for the target, because the only currrent would be the 10k, and as long as the serial adapter is works like it's supposed to (note: none of the ones I've played with recently do. DTR and RTS are NOT supposed to be push-pull, they are supposed to be internally pulled up and only forcefully able to drve low, /but on CH340 and the holtek chip (which also has the logic levels reported over USB for the 4 modem control inputs backwards; a pity since I both love modem control lines, and had been searching for a chip with the exact featureset that it has... have a pretty satisfying board design, except for the fact that the modem control line behave wrong... they start off both showing as not asserted, first edge it sees on one,  they all flip to inverted values.If they were all inverted all the time I might have continues on that design... )

3 pin UPDI, I put Gnd in the middle, which is also safe.

**Comment by Dlloydev** [comment]
Interesting discussion. I haven't tested the pyupdi UPDI programming method yet, but it should be quite easy to get the befefit of using the Shockley diode while still offering protection against contention (UPDI low while TX is high) even while connecting directly to the UPDI pin. How about something like this?

![image](https://user-images.githubusercontent.com/63488701/113888167-164a8780-9790-11eb-8937-ee6e41c7b17c.png)

**Comment by microPaul** [comment]
With this configuration the TX out of the USB-UART bridge can't pull the UPDI pin low.  Also, in a 5 volt system when the UPDI drives its output low it would be sinking about 10 mA, into the UPDI pin (assuming that the USB-UART can source 10mA from the TX output), which is quite a lot for a communications case.

**Reply by Dlloydev** [reply]
Whoops ... here's the corrected circuit.
 
**EDIT: Incorrect circuit removed**

So I guess the resistor in parallel with the diode should be sized as low as possible without exceeding the drive capability of the serial TX line and this also protects against the remote possibility of signal contention if UPDI goes high while TX is low.

**Comment by SpenceKonde** [comment]
Diode is backwards, idle serial line is HIGH. TX needs to be able to drive
the pin LOW

____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore: Arduino support for almost every ATTiny microcontroller
Contact: ***@***.***

**Comment by Dlloydev** [comment]
Ugh, I think I'll say away from anything electrical (or electronic) today!

![image](https://user-images.githubusercontent.com/63488701/113900577-f6b95c00-979b-11eb-99ca-f4ff9a99d560.png)

**Comment by SpenceKonde** [comment]
Uhhh.. nope,.... Combination of those...

What you want is a lower-resistance (say 470 ohm) resistor between target
and everything. From the adapter-facing side of that resistor, direct
connection to RX, and TX via schottky resistor with band towards the TX
pin.

On Wed, Apr 7, 2021 at 12:24 PM Dlloydev ***@***.***> wrote:

> Ugh, I think I'll say away from anything electrical (or electronic) today!
>
> [image: image]
> <https://user-images.githubusercontent.com/63488701/113900577-f6b95c00-979b-11eb-99ca-f4ff9a99d560.png>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/SpenceKonde/DxCore/discussions/94#discussioncomment-580771>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTXEW6NUQP2VCVJTURWBPTTHSBK7ANCNFSM42ALDBWQ>
> .
>


-- 
____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore <https://github.com/SpenceKonde/ATTinyCore>: Arduino support for
all pre-2016 tinyAVR with >2k flash!
megaTinyCore <https://github.com/SpenceKonde/megaTinyCore>: Arduino support
for all post-2016 tinyAVR parts!
DxCore <https://github.com/SpenceKonde/DxCore>: Arduino support for the AVR
Dx-series parts, the latest and greatest from Microchip!
Contact: ***@***.***

**Reply by Dlloydev** [reply]
Yes, I know, not shown because depending on the target board, could be anything from direct connect to 4.7K series resistor built-in. I wonder how crucial the extra series resistor is, because it would only protect the weakest link which is probably the inexpensive programmer with lower drive strength. However, this protection would be for a very remote possibility of signal contention.

When developing the jtag2updi HV programmers, I found a similar issue with the 4.7K series resistor, where it works, but there's "lifting" of the signal at higher speeds that can cause issues at higher baudrates, so I ended up protecting the programmer from contention and the HV pulse with a 1.2K series resistor and Shockley diode to VCC.

Note that on the Power Debugger, the signals are quite strong and buffered. The HV pulse is ultra fast and sharp (far beyond their published recommendation of 100µs min width as its only 30µs) and it is directly multiplexed onto the target UPDI signal.    

Back to speaking of the possibility of contention, on the [Arduino Due](https://www.arduino.cc/en/Hacking/PinMappingSAM3X), board, there are several shorted pins (PA29 <--> PC26) and (PA28 <--> PC29).

Note that the circuit as shown does have one-way protection due to the diode ... full protection if connected to one of your boards.
